helpers/embeds.js

```diff
@@ -0,0 +1,65 @@
+import { html, nothing, render } from 'lit';
+import { ifDefined } from 'lit/directives/if-defined.js';
+import { map } from 'lit/directives/map.js';
+import { tryGetPluginByKey } from '@brightspace-ui/core/helpers/plugins.js';
+
+const embedTypeAttributeName = 'data-d2l-embed-type';
+
+export async function createEmbedPlaceholder(embedType, options) {
+	const embedRendererPlugin = tryGetPluginByKey('d2l-html-embed-renderer', embedType);
+	if (!embedRendererPlugin) return;
+
+	const placeholderData = await embedRendererPlugin.getPlaceholder(options);
+	const contents = placeholderData.contents
+		? map(Object.entries(placeholderData.contents), ([id, content]) => html`<template data-d2l-embed-template-id=${id}>${content}</template>`)
+		: nothing;
+
+	const attributes = placeholderData.attributes ? JSON.stringify(placeholderData.attributes) : undefined;
+
+	return placeholderData.inline
+		? html`
+			<span data-d2l-embed-type="${embedType}" data-d2l-embed-props="${ifDefined(attributes)}">
+				${contents}
+			</span>
+		` : html`
+			<div data-d2l-embed-type="${embedType}" data-d2l-embed-props="${ifDefined(attributes)}">
+				${contents}
+			</div>
+		`;
+}
+
+export async function renderEmbeds(node, options) {
+	const placeholders = [...node.querySelectorAll(`div[${embedTypeAttributeName}], span[${embedTypeAttributeName}]`)];
+	if (placeholders.length === 0) return;
+
+	const processPlaceholder = async placeholder => {
+		const embedRendererPlugin = tryGetPluginByKey('d2l-html-embed-renderer', placeholder.dataset.d2lEmbedType);
+		if (!embedRendererPlugin) return;
+
+		const templates = [...placeholder.querySelectorAll('template[data-d2l-embed-template-id]')];
+		const processedTemplates = await Promise.all(templates.map(async template => {
+			const templateNode = template.content.cloneNode(true);
+			await renderEmbeds(templateNode, options);
+			return html`${templateNode}`;
```

Comment:
While `templateNode` is a document fragment, Lit nonetheless seems to handle this completely fine. I put together a test with a template that included a text node, some vanilla HTML, and a couple placeholders, and they all rendered into the DOM correctly.

Reply:
Yeah, the `html` template literal can handle a DOM node. In fact, I _think_ it handles arrays of nodes `${[nodes]}` just fine too.

Is it necessary to wrap `templateNode` in a Lit template here? It appears that it will result in `processedTemplateContents` being an array of Lit templates, which gets passed to `renderView(...)`. Does `renderView` need those to be Lit templates?

I checked out your user profile plugin to see how it would be used, but it doesn't appear to make use of this. I guess it's only for the nested case?

Reply:
Yeah, I deliberately want plugins to _only_ work with Lit templates. So they get Lit templates passed in via the plugin when there's nested content. The problem with passing raw DOM nodes around is that plugins can start walking the DOM to edit things they shouldn't.

While that can technically still happening (nothing stops a plugin from rendering these into a temporary DOM node and having at it), I'm hopeful providing what's effectively an opaque string will be a hint that they _shouldn't_ do this.